File: scrimage-core/src/main/scala/com/sksamuel/scrimage/Pixel.scala
package com.sksamuel.scrimage

import scala.language.implicitConversions

/**
 * A pixel is an encoding of a color value used in rasters.
 * The pixel is encoded using an ARGB packed int
 */
case class Pixel(argb: Int) {

  def alpha: Int = argb >> 24 & 0xFF
  def red: Int = argb >> 16 & 0xFF
  def green: Int = argb >> 8 & 0xFF
  def blue: Int = argb & 0xFF

  def toInt: Int = toARGBInt
  def toARGBInt: Int = argb
  def toColor: RGBColor = RGBColor(red, green, blue, alpha)

  def mapByComponent(f: Int => Int): Pixel = Pixel(f(red), f(green), f(blue), alpha)
}

object Pixel {

  implicit def int2pixel(pixel: Int): Pixel = new Pixel(pixel)

  def apply(r: Int, g: Int, b: Int, alpha: Int): Pixel = {
    val int = alpha << 24 | (r & 0xFF) << 16 | (g & 0xFF) << 8 | (b & 0xFF) << 0
    Pixel(int)
  }

  def apply(color: Color): Pixel = color.toPixel
}

File: scrimage-core/src/main/java/com/sksamuel/scrimage/Pixel.java
package com.sksamuel.scrimage;

import com.sksamuel.scrimage.color.RGBColor;

import java.util.function.Function;

/**
 * A pixel is an encoding of a color value used in rasters.
 * The pixel is encoded using an ARGB packed int
 */
public class Pixel {

    private final int argb;

    public Pixel(int argb) {
        this.argb = argb;
    }

    public Pixel(int r, int g, int b, int alpha) {
        this.argb = alpha << 24 | (r & 0xFF) << 16 | (g & 0xFF) << 8 | (b & 0xFF);
    }

    public int alpha() {
        return argb >> 24 & 0xFF;
    }

    public int red() {
        return argb >> 16 & 0xFF;
    }

    public int green() {
        return argb >> 8 & 0xFF;
    }

    public int blue() {
        return argb & 0xFF;
    }

    public int toInt() {
        return toARGBInt();
    }

    public int toARGBInt() {
        return argb;
    }

    public RGBColor toColor() {
        return new RGBColor(red(), green(), blue(), alpha());
    }

    public Pixel mapByComponent(Function<Integer, Integer> f) {
        return new Pixel(f.apply(red()), f.apply(blue()), f.apply(green()), f.apply(alpha()));
    }
}